File: app/src/main/java/org/jellyfin/androidtv/ui/ClockUserView.java
package org.jellyfin.androidtv.ui;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextClock;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import org.jellyfin.androidtv.R;
import org.jellyfin.androidtv.TvApp;
import org.jellyfin.androidtv.preference.UserPreferences;
import org.jellyfin.androidtv.preference.constant.ClockBehavior;
import org.jellyfin.androidtv.ui.playback.PlaybackOverlayActivity;
import org.jellyfin.androidtv.util.ImageUtils;
import org.jellyfin.apiclient.interaction.ApiClient;

import static org.koin.java.KoinJavaComponent.get;

public class ClockUserView extends RelativeLayout {
    public ClockUserView(Context context) {
        super(context);
        init(context);
    }

    public ClockUserView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View v = inflater.inflate(R.layout.clock_user_bug, null, false);
        this.addView(v);
        TextClock tc = v.findViewById(R.id.clock);
        ClockBehavior showClock = get(UserPreferences.class).get(UserPreferences.Companion.getClockBehavior());
        Activity activity;
        switch (showClock) {
            case ALWAYS:
                tc.setVisibility(VISIBLE);
                break;
            case NEVER:
                tc.setVisibility(GONE);
                break;
            case IN_VIDEO:
                activity = getActivity();
                if (!(activity instanceof PlaybackOverlayActivity))
                    tc.setVisibility(GONE);
                else
                    tc.setVisibility(VISIBLE);
                break;
            case IN_MENUS:
                activity = getActivity();
                if (activity instanceof PlaybackOverlayActivity)
                    tc.setVisibility(GONE);
                else
                    tc.setVisibility(VISIBLE);
                break;
            default:
                break;
        }
        if (!isInEditMode()) {
            TextView username = ((TextView) v.findViewById(R.id.userName));
            username.setText(TvApp.getApplication().getCurrentUser().getName());
            ImageView userImage = (ImageView) v.findViewById(R.id.userImage);
            if (TvApp.getApplication().getCurrentUser().getPrimaryImageTag() != null) {
                Glide.with(context)
                        .load(ImageUtils.getPrimaryImageUrl(TvApp.getApplication().getCurrentUser(), get(ApiClient.class)))
                        .error(R.drawable.ic_user)
                        .override(30, 30)
                        .centerInside()
                        .into(userImage);
            } else {
                userImage.setImageResource(R.drawable.ic_user);
            }

        }
    }

    private Activity getActivity() {
        Context context = getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof Activity) {
                return (Activity) context;
            }
            context = ((ContextWrapper) context).getBaseContext();
        }
        return null;
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/ClockUserView.kt
package org.jellyfin.androidtv.ui

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import org.jellyfin.androidtv.R
import org.jellyfin.androidtv.auth.repository.UserRepository
import org.jellyfin.androidtv.databinding.ClockUserBugBinding
import org.jellyfin.androidtv.preference.UserPreferences
import org.jellyfin.androidtv.preference.constant.ClockBehavior
import org.jellyfin.androidtv.ui.navigation.Destinations
import org.jellyfin.androidtv.ui.navigation.NavigationRepository
import org.jellyfin.androidtv.util.ImageHelper
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class ClockUserView @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
	defStyleAttr: Int = 0,
	defStyleRes: Int = 0,
) : RelativeLayout(context, attrs, defStyleAttr, defStyleRes), KoinComponent {
	private val binding: ClockUserBugBinding = ClockUserBugBinding.inflate(LayoutInflater.from(context), this, true)
	private val userPreferences by inject<UserPreferences>()
	private val userRepository by inject<UserRepository>()
	private val navigationRepository by inject<NavigationRepository>()
	private val imageHelper by inject<ImageHelper>()

	var isVideoPlayer = false
		set(value) {
			field = value
			updateClockVisibility()
		}

	val homeButton get() = binding.home

	init {
		updateClockVisibility()

		val currentUser = userRepository.currentUser.value

		binding.clockUserImage.load(
			url = currentUser?.let(imageHelper::getPrimaryImageUrl),
			placeholder = ContextCompat.getDrawable(context, R.drawable.ic_user)
		)

		binding.clockUserImage.isVisible = currentUser != null

		binding.home.setOnClickListener {
			navigationRepository.reset(Destinations.home, clearHistory = true)
		}
	}

	private fun updateClockVisibility() {
		val showClock = userPreferences[UserPreferences.clockBehavior]

		binding.clock.isVisible = when (showClock) {
			ClockBehavior.ALWAYS -> true
			ClockBehavior.NEVER -> false
			ClockBehavior.IN_VIDEO -> isVideoPlayer
			ClockBehavior.IN_MENUS -> !isVideoPlayer
		}

		binding.home.isVisible = !isVideoPlayer
	}
}
